### PR TITLE
Streamed active users cruncher

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

File: crunchers/cruncher.h [hunk 1]
```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
 using StreamCruncher =
     current::ss::StreamSubscriber<GenericCruncherImpl<IMPL, BUFFER_SIZE>, typename IMPL::event_t>;
 
+// An extension of the `GenerucCruncherImpl` wrapper above,
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: Typo: Gener**i**cCruncherImpl

File: crunchers/cruncher.h [hunk 2]
```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
...
+// which also generates "ticks" at specified time intervals.
+// The IMPL has the same requirements except it must also implement `OnTick()` method.
+template <typename IMPL, size_t BUFFER_SIZE>
+struct GenericTickCruncherImpl : public GenericCruncherImpl<IMPL, BUFFER_SIZE> {
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: - s/GenericTickCruncherImpl/GenericCruncherWithTicksImpl/?
- Public inheritance is default for structs. Maybe you want to make them classes?

File: crunchers/cruncher.h [hunk 3]
```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
...
+    }
+    last_event_us_ = us;
+  }
+
+  using base_t::mmq_;
+  const std::chrono::microseconds tick_interval_;
+  std::chrono::microseconds last_event_us_;
+};
+
+template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: Add some global constant `kDefaultMMQBufferSize`?

File: crunchers/cruncher.h [hunk 4]
```diff
@@ -200,11 +266,72 @@ class MultiCruncher {
...
+  using base_t = MultiCruncher<CRUNCHER>;
+  using event_t = typename base_t::event_t;
+  using response_t = typename base_t::response_short_t;
+  using params_list_t = std::vector<CruncherOutputStreamParams>;
+  using streams_list_t = std::vector<OUTPUT_STREAM>;
+
+  template <typename ARG>
+  StreamedMultiCruncher(const std::vector<ARG>& args, const params_list_t& params)
+      : base_t(args), params_(params) {
+    assert(args.size() == params.size());
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: BTW, we have `CURRENT_ASSERT` macro.

File: crunchers/cruncher.h [hunk 5]
```diff
@@ -173,6 +232,13 @@ class MultiCruncher {
     }
   }
 
+  void OnTick(std::chrono::microseconds us) {
+    last_event_us_ = us;
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: `last_event_us_` -> `last_event_or_tick_us_`?

File: crunchers/cruncher.h [hunk 6]
```diff
@@ -200,11 +266,72 @@ class MultiCruncher {
...
+        crunchers_[i]->OnTick(response.timestamp);
+        response.value = crunchers_[i]->GetValue();
+        streams_[i].Publish(response);
+      }
+    }
+    base_t::OnTick(us);
+  }
+
+  void OnEvent(event_t&& event, idxts_t idxts) {
+    OnTick(idxts.us);
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: Calling `OnTick()` here is a bit confusing. Maybe it's better to move the code above to some `GenerateTicks()` method?

File: crunchers/cruncher.h [hunk 7]
```diff
@@ -200,11 +266,72 @@ class MultiCruncher {
...
+    assert(args.size() == params.size());
+    streams_.reserve(params.size());
+    for (const auto& param : params) {
+      streams_.emplace_back(param.path);
+    }
+  }
+  virtual ~StreamedMultiCruncher() = default;
+
+  void OnTick(std::chrono::microseconds us) {
+    std::chrono::microseconds last_us = last_event_us_;
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: const

File: crunchers/cruncher.h [hunk 8]
```diff
@@ -200,11 +266,72 @@ class MultiCruncher {
...
+    for (size_t i = 0, sz = streams_.size(); i < sz; ++i) {
+      const uint64_t publish_interval = params_[i].interval.count();
+      const uint64_t start_marker = last_us.count() / publish_interval;
+      const uint64_t end_marker = us.count() / publish_interval;
+      for (uint64_t j = start_marker; j < end_marker; ++j) {
+        response_t response;
+        response.timestamp = std::chrono::microseconds((j + 1) * publish_interval);
+        crunchers_[i]->OnTick(response.timestamp);
+        response.value = crunchers_[i]->GetValue();
+        streams_[i].Publish(response);
```
Comment (mzhurovich) [inline crunchers/cruncher.h]: If we work with historical data, I suppose we should use the second argument for `Publish()` to make this response properly timestamped. @dkorolev ?

File: crunchers/cruncher_active_users.h [hunk 9]
```diff
@@ -52,13 +49,25 @@ struct ActiveUsersCruncherImpl {
     if (Exists<EventLogEntry>(e)) {
       OnEventLogEntry(Value<EventLogEntry>(e));
     }
-    while (!users_list_.empty() && users_list_.back().us + interval_ <= current_us_) {
+    OnTick(idxts.us);
```
Comment (mzhurovich) [inline crunchers/cruncher_active_users.h]: Same here. Calling `OnTick` from `OnEvent` is not obvious for the reader.

File: crunchers/cruncher.h [hunk 10]
```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
...
+// The IMPL has the same requirements except it must also implement `OnTick()` method.
+template <typename IMPL, size_t BUFFER_SIZE>
+struct GenericTickCruncherImpl : public GenericCruncherImpl<IMPL, BUFFER_SIZE> {
+  using base_t = GenericCruncherImpl<IMPL, BUFFER_SIZE>;
+  using event_t = typename base_t::event_t;
+  using EntryResponse = current::ss::EntryResponse;
+
+  class TickMessage final : public base_t::Message {
+   public:
+    TickMessage(const std::chrono::microseconds& timestamp) : timestamp_(timestamp) {}
```
Comment (dkorolev) [inline crunchers/cruncher.h]: `std::chrono::microseconds` is safe to pass by value :)

File: crunchers/cruncher.h [hunk 11]
```diff
@@ -136,6 +138,63 @@ template <typename IMPL, size_t BUFFER_SIZE = 1024 * 1024>
...
+    GenerateTickEvents(current.us);
+    return base_t::operator()(std::move(event), current, last);
+  }
+
+ private:
+  void GenerateTickEvents(std::chrono::microseconds us) {
+    const uint64_t start_marker = last_event_us_.count() / tick_interval_.count();
+    const uint64_t end_marker = us.count() / tick_interval_.count();
+    for (uint64_t i = start_marker; i < end_marker; ++i) {
+      mmq_.Publish(std::make_unique<TickMessage>(std::chrono::microseconds((i + 1) * tick_interval_.count())));
```
Comment (dkorolev) [inline crunchers/cruncher.h]: Why not a loop from `std::chrono::microseconds(((last_event_us_.count() / tick_interval_.count()) + 1) * tick_interval_.count())` to `<= us.count()` and `+= tick_interval_count_` ?

File: crunchers/cruncher.h (hunk 5 above)
Comment (dkorolev) [reply crunchers/cruncher.h]: +1

File: crunchers/cruncher.h (hunk 4 above)
Comment (dkorolev) [reply crunchers/cruncher.h]: +1

File: crunchers/cruncher.h (hunk 8 above)
Comment (dkorolev) [reply crunchers/cruncher.h]: +1

**Review by dkorolev — APPROVED** [review]
+1 to Max's comments and a few of mine.

`OnTick()` called manually is indeed confusing, and it's best to move its logic into a custom function.

File: crunchers/cruncher.h (hunk 5 above)
Comment (grixa) [reply crunchers/cruncher.h]: May be just `last_us`? :)

File: crunchers/cruncher.h (hunk 10 above)
Comment (grixa) [reply crunchers/cruncher.h]: :)

File: crunchers/cruncher.h [hunk 12]
```diff
@@ -178,10 +186,11 @@ struct GenericTickCruncherImpl : public GenericCruncherImpl<IMPL, BUFFER_SIZE> {
 
  private:
   void GenerateTickEvents(std::chrono::microseconds us) {
-    const uint64_t start_marker = last_event_us_.count() / tick_interval_.count();
-    const uint64_t end_marker = us.count() / tick_interval_.count();
-    for (uint64_t i = start_marker; i < end_marker; ++i) {
-      mmq_.Publish(std::make_unique<TickMessage>(std::chrono::microseconds((i + 1) * tick_interval_.count())));
+    for (auto t = std::chrono::microseconds(((last_event_us_.count() / tick_interval_.count()) + 1) *
```
Comment (dkorolev) [inline crunchers/cruncher.h:189]: :+1:

File: crunchers/cruncher.h [hunk 13]
```diff
@@ -226,29 +235,29 @@ class MultiCruncher {
...
     for (auto& cruncher : crunchers_) {
       cruncher->OnTick(us);
     }
   }
 
   void OnRequest(Request&& r) {
-    const std::string comment =
-        Printf("Last processed event was %llu minutes ago",
-               std::chrono::duration_cast<std::chrono::minutes>(current::time::Now() - last_event_us_).count());
+    const std::string comment = Printf("Last processed event was %llu minutes ago",
```
Comment (dkorolev) [inline crunchers/cruncher.h]: I believe that thing is signed -- `%lld`?

File: crunchers/cruncher.h [hunk 14]
```diff
@@ -302,32 +311,39 @@ class StreamedMultiCruncher : public MultiCruncher<CRUNCHER> {
...
  private:
+  void OutputCrunchersDataUntil(std::chrono::microseconds us) {
+    const std::chrono::microseconds last_us = last_event_or_tick_us_;
+    for (size_t i = 0, sz = streams_.size(); i < sz; ++i) {
+      const auto publish_interval = params_[i].interval;
+      for (auto t = std::chrono::microseconds((last_us.count() / publish_interval.count() + 1) *
+                                              publish_interval.count());
+           t < us;
+           t += publish_interval) {
+        // The result of GetValue() call could be different if corresponding cruncher
```
Comment (dkorolev) [inline crunchers/cruncher.h]: The result of  **the** ... if **the** corresponding cruncher ... that **the** "actual" timestamp has changed. :)

File: crunchers/cruncher_active_users.h [hunk 15]
```diff
@@ -77,6 +74,13 @@ struct ActiveUsersCruncherImpl {
   using user_list_t = std::list<ActiveUser>;
   using user_map_t = std::unordered_map<std::string, typename user_list_t::iterator>;
 
+  void RemoveUsersOutsideTheTimeWindow() {
+    while (!users_list_.empty() && users_list_.back().us + interval_ <= current_us_) {
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h:78]: If the time window it STL style, `[ inclusive, exclusive )`, the last comparison should be `< current_us_`, not `<= current_us_`.

Comment (grixa) [reply crunchers/cruncher_active_users.h:78]: I don't quite understand what means the `[ inclusive, exclusive )` here. Do you suggest to exclude from the time window the last event, whose timestamp equals `current_us_`? If not, here is the problem: suppose we generate exactly one event per second and set the `interval_` to `std::chrono::seconds(1)`. Then, I suppose, we expect to see only one event inside the time window at any moment. And it won't be true if the condition is `...  < current_us_`

Comment (grixa) [reply crunchers/cruncher_active_users.h:78]: May be `< current_us_ + std::chrono::microseconds(1)` (as in the persisters) is a better solution.. What do you think?

Comment (mzhurovich) [reply crunchers/cruncher_active_users.h:78]: Imagine you have `time_window_size = std::chrono::microseconds(1)`. You publish two events with the timestamps equal to `0` and `1` respectively. How many events do you expect to see in your time window buffer?

Comment (grixa) [reply crunchers/cruncher_active_users.h:78]: It's the same example as the previous one with seconds. If we publish 1 event per microsecond, we expect to see only one event in 1-microsecond window, right?

Comment (mzhurovich) [reply crunchers/cruncher_active_users.h:78]: Sure. And the logic you've implemented seems to be correct ;)

**Comment by dkorolev** [comment]
:+1:

**Pull request merged**